### PR TITLE
Flow manager adaptive/v9

### PR DESCRIPTION
**Pull request by victorjulien**
#6590 with some adjustments for offline mode, some cleanups.

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | test | baseline | % |
|--|--|--|--| 
| | tlpw2_single_stats_chk |
| .flow.wrk.spare_sync | 11025 | 14544 | 75.8% |
| .flow.wrk.flows_evicted | 583024 | 72254 | 806.91% |
| .flow.wrk.flows_injected | 287055 | 320739 | 89.5% |
| .flow.mgr.full_hash_pass | 272 | 624 | 43.59% |
| .flow.mgr.flows_checked | 1480989 | 2991145 | 49.51% |
| .flow.mgr.flows_notimeout | 599727 | 1216594 | 49.3% |
| .flow.mgr.flows_timeout | 881262 | 1774551 | 49.66% |
| .flow.mgr.flows_evicted | 881262 | 1774551 | 49.66% |
| .flow.mgr.flows_evicted_needs_work | 161327 | 320739 | 50.3% |
| .flow.memuse | 550313728 | 454217728 | 121.16% |
| .tcp.synack | 902476 | 776957 | 116.16% |
| .tcp.rst | 1239571 | 1114052 | 111.27% |
| | tlpw2_autofp_stats_chk |
| .uptime | 255 | 229 | 111.35% |
| .flow.wrk.spare_sync | 11103 | 14534 | 76.39% |
| .flow.wrk.flows_evicted | 577454 | 74697 | 773.06% |
| .flow.wrk.flows_injected | 287097 | 321061 | 89.42% |
| .flow.mgr.full_hash_pass | 87 | 620 | 14.03% |
| .flow.mgr.flows_checked | 1070769 | 3004502 | 35.64% |
| .flow.mgr.flows_notimeout | 262557 | 1241275 | 21.15% |
| .flow.mgr.flows_timeout | 808212 | 1763227 | 45.84% |
| .flow.mgr.flows_evicted | 810365 | 1764303 | 45.93% |
| .flow.mgr.flows_evicted_needs_work | 161101 | 320626 | 50.25% |
| .flow.memuse | 582505728 | 454217728 | 128.24% |
| | tlpw1_stats_chk |
| .flow.wrk.spare_sync | 4535 | 5145 | 88.14% |
| .flow.wrk.flows_evicted | 117173 | 25236 | 464.31% |
| .flow.mgr.full_hash_pass | 99 | 9242 | 1.07% |
| .flow.mgr.flows_checked | 276722 | 783985 | 35.3% |
| .flow.mgr.flows_notimeout | 10695 | 197056 | 5.43% |
| .flow.mgr.flows_timeout | 266027 | 586929 | 45.33% |
| .flow.mgr.flows_evicted | 266027 | 586929 | 45.33% |
| .flow.mgr.flows_evicted_needs_work | 42108 | 84110 | 50.06% |
| .flow.memuse | 518793728 | 454217728 | 114.22% |
| .tcp.synack | 183413 | 157473 | 116.47% |
| .tcp.rst | 136174 | 89317 | 152.46% |
| | generic_stats_chk |
| .flow.wrk.flows_evicted_needs_work | 1828829 | 2115348 | 86.46% |
| .flow.wrk.flows_evicted_pkt_inject | 3504135 | 4052316 | 86.47% |
| .flow.wrk.flows_evicted | 64097 | 77611 | 82.59% |
| .flow.wrk.flows_injected | 1803834 | 2089610 | 86.32% |
| .flow.mgr.flows_checked | 3261214 | 2151610 | 151.57% |
| .flow.mgr.flows_notimeout | 1980715 | 859269 | 230.51% |
| .flow.mgr.flows_evicted | 1431282 | 1760131 | 81.32% |
| .flow.mgr.flows_evicted_needs_work | 1034941 | 1308834 | 79.07% |
| .flow.memuse | 1005097728 | 910313728 | 110.41% | 

Pipeline 4810

File: src/app-layer.c [hunk 1]
```diff
@@ -982,9 +1053,21 @@ void AppLayerRegisterThreadCounters(ThreadVars *tv)
...
+                        applayer_counter_names[ipproto_map][alproto].alloc_error, tv);
+                applayer_counters[ipproto_map][alproto].parser_error_id = StatsRegisterCounter(
+                        applayer_counter_names[ipproto_map][alproto].parser_error, tv);
+                applayer_counters[ipproto_map][alproto].internal_error_id = StatsRegisterCounter(
+                        applayer_counter_names[ipproto_map][alproto].internal_error, tv);
             } else if (alproto == ALPROTO_FAILED) {
                 applayer_counters[ipproto_map][alproto].counter_id =
                     StatsRegisterCounter(applayer_counter_names[ipproto_map][alproto].name, tv);
+
+                applayer_counters[ipproto_map][alproto].gap_error_id = StatsRegisterCounter(
```
Comment (victorjulien) [inline src/app-layer.c:1069]: only register for tcp

File: src/app-layer.c [hunk 2]
```diff
@@ -982,9 +1053,21 @@ void AppLayerRegisterThreadCounters(ThreadVars *tv)
 
                 applayer_counters[ipproto_map][alproto].counter_tx_id =
                     StatsRegisterCounter(applayer_counter_names[ipproto_map][alproto].tx_name, tv);
+
+                applayer_counters[ipproto_map][alproto].gap_error_id = StatsRegisterCounter(
```
Comment (victorjulien) [inline src/app-layer.c:1057]: only register for tcp

File: src/flow-manager.c [hunk 3]
```diff
@@ -736,6 +758,42 @@ static uint32_t FlowTimeoutsMin(void)
...
+    full_pass_in_ms *= perc;
+    full_pass_in_ms = MAX(full_pass_in_ms, 333);
+
+    uint32_t work_unit_ms = 999 * perc;
+    work_unit_ms = MAX(work_unit_ms, 250);
+
+    uint32_t wus_per_full_pass = full_pass_in_ms / work_unit_ms;
+
+    uint32_t rows_per_wu = MAX(1, rows / wus_per_full_pass);
+    uint32_t rows_process_cost = rows_per_wu / 1000; // est 1usec per row
```
Comment (inashivb) [inline src/flow-manager.c:766]: So, if we have up to 999 rows per worker units, the processing cost will be 0 and not 999usec ~ 1ms?

File: src/flow-manager.c [hunk 4]
```diff
@@ -736,6 +758,42 @@ static uint32_t FlowTimeoutsMin(void)
...
+    if (emergency) {
+        *wu_rows = rows;
+        *wu_sleep = 250;
+        return;
+    }
+
+    uint32_t full_pass_in_ms = pass_in_sec * 1000;
+    float perc = MIN((((float)(100 - mp) / (float)100)), 1);
+    full_pass_in_ms *= perc;
+    full_pass_in_ms = MAX(full_pass_in_ms, 333);
```
Comment (inashivb) [inline src/flow-manager.c:758]: Q: What does 333 mean?

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/6874

**Pull request closed without merging**